### PR TITLE
Dashboard backend

### PR DESCRIPTION
**Pull request by TheKinrar**

### CODE REVIEW

File: src/main/scala/services/AuthService.scala [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import cats.effect.std.UUIDGen
+
+import java.util.UUID
+import scala.util.Random
+
+class AuthService(repos: Repositories) {
+  given UUIDGen[IO] = UUIDGen.fromSync
+
+  private def generateStaticToken(): IO[String] =
+    IO(Random.alphanumeric.take(128).mkString)
```
Comment (m4rk055) [inline src/main/scala/services/AuthService.scala]: Use `SecureRandom`

File: build.sbt [hunk 2]
```diff
@@ -15,17 +15,21 @@ lazy val root = (project in file("."))
...
       "com.softwaremill.sttp.tapir" %% "tapir-swagger-ui-bundle"         % tapirVersion,
+      "io.blindnet"                 %% "identity-client"                 % "1.0.1-SNAPSHOT",
       "io.circe"                    %% "circe-core"                      % circeVersion,
       "io.circe"                    %% "circe-generic"                   % circeVersion,
       "io.circe"                    %% "circe-literal"                   % circeVersion % Test,
+      "org.apache.commons"          %  "commons-email"                   % "1.5",
       "org.flywaydb"                %  "flyway-core"                     % "9.3.0",
       "org.http4s"                  %% "http4s-blaze-server"             % http4sVersion,
       "org.http4s"                  %% "http4s-circe"                    % http4sVersion,
+      "org.mindrot"                 %  "jbcrypt"                         % "0.4",
```
Comment (m4rk055) [inline build.sbt]: This lib hasn't been update for [8 years](https://github.com/jeremyh/jBCrypt).
[password4j](https://github.com/Password4j/password4j) looks like a good alternative.

File: src/main/resources/db/migration/V5__Account_verification.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+alter table accounts
+    add verified boolean not null default false,
+    add email_token text not null default '';
```
Comment (m4rk055) [inline src/main/resources/db/migration/V5__Account_verification.sql:3]: Add creation date so we can easily remind users to verify (and handle various legal actions).

File: src/main/scala/util/Resources.scala [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import cats.data.OptionT
+import cats.effect.*
+
+object Resources {
+  def getAsStream(resource: String): IO[Option[fs2.Stream[IO, Byte]]] =
+    IO(Option(getClass.getResourceAsStream(resource)))
+      .map(_.map(s => fs2.io.readInputStream(IO.pure(s), 1000)))
+
+  def readAsString(resource: String): IO[Option[String]] =
```
Comment (m4rk055) [inline src/main/scala/util/Resources.scala:12]: fs2 has [the file api](https://fs2.io/#/io?id=files)

File: src/main/scala/endpoints/AuthEndpoints.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+  val status: ApiEndpoint =
+    authedBase.summary("Get account status")
+      .get
+      .in("status")
+      .out(jsonBody[AccountStatusPayload])
+      .serverLogicSuccess(service.status)
+
+  val verify: ApiEndpoint =
+    authedBase.summary("Verify email")
```
Comment (m4rk055) [inline src/main/scala/endpoints/AuthEndpoints.scala]: For UX, I think this should be a public endpoint, as users expect to have their acc verified as soon as they open the link they received in the email msg (not having to log in first).

File: src/main/scala/models/Account.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+case class Account(id: UUID, email: String, password: String, token: String, verified: Boolean, emailToken: String) extends St {
+  def verifyPassword(in: String): IO[Boolean] = IO {
+    BCrypt.checkpw(in, password)
+  }
+}
+
+object Account {
+  def hashPassword(password: String): IO[String] = IO {
+    BCrypt.hashpw(password, BCrypt.gensalt())
```
Comment (m4rk055) [inline src/main/scala/models/Account.scala]: Shouldn't salt be stored? Or is it concatenated to the password hash?

File: src/main/scala/services/AuthService.scala [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+  private def sendVerificationEmail(email: String, emailToken: String): IO[Unit] =
+    def processTemplate(template: String, env: Env): String =
+      template.replace("$verification_link$", s"${env.baseUrl}/verify?token=$emailToken")
+
+    for {
+      env <- Env.get
```
Comment (m4rk055) [inline src/main/scala/services/AuthService.scala]: I prefer reading the config once and passing the object as a parameter. Makes it easier to debug/test

File: src/main/scala/services/AuthService.scala [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    for {
+      _ <- (!acc.verified).orBadRequest("Already verified")
+      _ <- sendVerificationEmail(acc.email, acc.emailToken)
+    } yield ()
+
+  def updateEmail(acc: Account)(payload: UpdateEmailPayload): IO[Unit] =
+    for {
+      _ <- acc.verifyPassword(payload.current_password).map(_.orForbidden)
+      emailToken <- generateStaticToken()
+      _ <- repos.accounts.updateEmail(acc.id, payload.email, emailToken)
```
Comment (m4rk055) [inline src/main/scala/services/AuthService.scala:68]: Also send verification to the new email address.

File: src/main/scala/mail/Mail.scala [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import scala.util.Try
+
+case class Mail(
+  from: Sender,
+  to: List[String],
+  subject: String,
+  message: String,
+  richMessage: Option[String] = None,
+) {
+  def send(config: MailConfig): IO[Unit] = IO {
```
Comment (m4rk055) [inline src/main/scala/mail/Mail.scala]: By fp guidelines, `case class` shouldn't really cause any effects nor depend on services. It is just a simple [ADT](https://blog.rockthejvm.com/algebraic-data-types/) representing a part of the domain.
Change it into `class` or make the `EmailService` (I prefer the latter).

**Review by m4rk055 — CHANGES_REQUESTED** [review]

File: src/main/scala/util/Resources.scala (hunk 4 above)
Comment (TheKinrar) [reply src/main/scala/util/Resources.scala:12]: I did not find a function for reading from a resource (but could have missed it)

File: src/main/scala/models/Account.scala (hunk 6 above)
Comment (TheKinrar) [reply src/main/scala/models/Account.scala]: It's the typical bcrypt format with salt and parameters in the output, yes.

File: src/main/scala/util/Resources.scala (hunk 4 above)
Comment (m4rk055) [reply src/main/scala/util/Resources.scala:12]: Right. [fs2.io.readClassResource](https://github.com/typelevel/fs2/blob/d306fe880890d4b76ab7d69e89e8365e17e75b79/io/jvm-native/src/main/scala/fs2/io/iojvmnative.scala#L68) seems to do exactly the same as you did.
There is also this [PR](https://github.com/typelevel/fs2/pull/3073/files) from a couple of days ago, to use the file api with in-memory file system, but is more complicated.

**Review by m4rk055 — APPROVED** [review]

**Pull request merged**